crates/chain/src/keychain/txout_index.rs

```diff
@@ -13,6 +13,76 @@ use core::{
 
 use crate::Append;
 
+/// Represents updates to the derivation index of a [`KeychainTxOutIndex`].
+/// It maps each keychain `K` to a descriptor and its last revealed index.
+///
+/// It can be applied to [`KeychainTxOutIndex`] with [`apply_changeset`]. [`ChangeSet] are
+/// monotone in that they will never decrease the revealed derivation index.
+///
+/// [`KeychainTxOutIndex`]: crate::keychain::KeychainTxOutIndex
+/// [`apply_changeset`]: crate::keychain::KeychainTxOutIndex::apply_changeset
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "K: Ord + serde::Deserialize<'de>",
+            serialize = "K: Ord + serde::Serialize"
+        )
+    )
+)]
+#[must_use]
+pub struct ChangeSet<K> {
+    /// Contains the keychains that have been added and their respective descriptor
+    pub keychains_added: BTreeMap<K, Descriptor<DescriptorPublicKey>>,
```

Comment:
I've run into a problem testing this PR with the `wallet_esplora_async` test. When it reloads this `ChangeSet` since `keychains_added` is a `Descriptor<DescriptorPublicKey>>` I'm not able to sign any transactions. Am I supposed to manually add privkey signers to my `Wallet` or something like that ? Or do we need a way to store the private key descriptors in the db?

Reply:
Definitely not store privkey in db imo. Maybe have an option to load privkey when constructing wallet.

Reply:
This is a good point. I don't think we considered the impact on private descriptors to `Wallet` when designing this change which was an oversight. Yes you're meant to manually add them and I expect this is quite difficult. In the long term we should be looking to remove the signers from `Wallet` but we probably need a design that makes it easy to add the keys from private descriptors to the Wallet for now before merging this.

Reply:
Ok, it's not too hard to manually add the signers, but will need to update the examples and docs. 

We could make this easier by adding a function to add a whole `SignerContainer` or in `Wallet::new_or_load` add the signers if a `Descriptor<DescriptorSecretKey>` is passed in. Here's my quick examples workaround:

```rust
    let secp = Secp256k1::new();

    let (external_descriptor, external_keymap) = Descriptor::parse_descriptor(&secp, "wpkh(tprv8ZgxMBicQKsPdy6LMhUtFHAgpocR8GC6QmwMSFpZs7h6Eziw3SpThFfczTDh5rW2krkqffa11UpX3XkeTTB2FvzZKWXqPY54Y6Rq4AQ5R8L/84'/1'/0'/0/*)").unwrap();
    let (internal_descriptor, internal_keymap) = Descriptor::parse_descriptor(&secp, "wpkh(tprv8ZgxMBicQKsPdy6LMhUtFHAgpocR8GC6QmwMSFpZs7h6Eziw3SpThFfczTDh5rW2krkqffa11UpX3XkeTTB2FvzZKWXqPY54Y6Rq4AQ5R8L/84'/1'/0'/1/*)").unwrap();

    let external_signer_container = SignersContainer::build(external_keymap, &external_descriptor, &secp);
    let internal_signer_container = SignersContainer::build(internal_keymap, &internal_descriptor, &secp);
    
    let mut wallet = Wallet::new_or_load(
        external_descriptor,
        Some(internal_descriptor),
        db,
        Signet,
    )?;
    
    external_signer_container.signers().into_iter()
        .for_each(|s| wallet.add_signer(KeychainKind::External, SignerOrdering::default(), s.clone()));
    internal_signer_container.signers().into_iter()
        .for_each(|s| wallet.add_signer(KeychainKind::Internal, SignerOrdering::default(), s.clone()));
```

Reply:
I added code in the wallet to add the expected descriptor keys as signers and I think it's a pretty clean way to do it: https://github.com/bitcoindevkit/bdk/pull/1128/commits/c053f1940013987bdf0902c41f074158eaafabad.